tianshou/data/utils/segtree.py

```diff
@@ -0,0 +1,117 @@
+import numpy as np
+from typing import Union, Optional
+# from numba import njit
+
+
+class SegmentTree:
+    """Implementation of Segment Tree. The procedure is as follows:
+
+    1. Find out the smallest n which safisfies ``size <= 2^n``, and let \
+    ``bound = 2^n``. This is to ensure that all leaf nodes are in the same \
+    depth inside the segment tree.
+    2. Store the original value to leaf nodes in ``[bound:bound * 2]``, and \
+    the union of elementary to internal nodes in ``[1:bound]``. The internal \
+    node follows the rule: \
+    ``value[i] = operation(value[i * 2], value[i * 2 + 1])``.
+    3. Update a node takes O(log(bound)) time complexity.
+    4. Query an interval [l, r] with the default operation takes O(log(bound))
```

Comment:
I don't think these implementation details are necessary in the doc of SegmentTree. Padding the array to size of 2^n is just an implementation detail.

A more proper way is to say something like the following:

Store an array ``arr`` with size ``n`` in a segment tree, support value update and fast query of ``min/max/sum arr[left:right]`` in O(log(n)) time.